docs/proposals/scheduling/20220629-fine-grained-device-scheduling.md

```diff
@@ -0,0 +1,419 @@
+
+### Non-goals/Future work
+
+1. Define flexible allocation strategies, such as implementing BinPacking or Spread according to GPU resources
+
+## Proposal
+
+### API
+
+#### Device resource dimensions
+
+Due to GPU is complicated, we will introduce GPU first. As we all know there is compute and GPU Memory capability for the GPU device. Generally user apply GPU like "I want 1/2/4/8 GPUs", but if node support GPU level isolation mechanism, user may apply GPU like "I want 0.5/0.25 GPU resources". Moreover, user may set different compute capability and GPU memory capability for best resource utilization, so the user want apply GPU like "I want X percent of "compute capability and Y percent of memory capability".
+
+We abstract GPU resources into different dimensions:
+
+- `koordinator.sh/gpu-core` represents the computing capacity of the GPU. Similar to K8s MilliCPU, we abstract the total computing power of GPU into one hundred, and users can apply for the corresponding amount of GPU computing power according to their needs.
+- `koordinator.sh/gpu-memory` represents the memory capacity of the GPU in bytes.
+- `koordinator.sh/gpu-memory-ratio` represents the percentage of the GPU's memory.
+
+Assuming that node A has 4 GPU instances, and the total memory of each instance is 8GB, when device reporter reports GPU capacity information to `Node.Status.Allocatable`, it no longer reports nvidia.com/gpu=4, but reports the following information:
+
+```yaml
+status:
+  capacity:
+    koordinator.sh/gpu-core: 400
+    koordinator.sh/gpu-memory: "32GB"
+    koordinator.sh/gpu-memory-ratio: 400
+  allocatable:
+    koordinator.sh/gpu-core: 400
+    koordinator.sh/gpu-memory: "32GB"
+    koordinator.sh/gpu-memory-ratio: 400
+```
+
+For the convenience of users, an independent resource name `koordinator.sh/gpu` is defined. For example, when a user wants to use half of the computing resources and memory resources of a GPU instance, the user can directly declare `koordinator.sh/gpu: 50`, and the scheduler will convert it to `koordinator.sh/gpu-core: 50, koordinator.sh/gpu-memory-ratio: 50`
+
+For other devices like RDMA and FPGA, the node has 1 RDMA and 1 FGPA, will report the following information:
+
+```yaml
+status:
+  capacity:
+    koordinator.sh/rdma: 100
+    koordinator.sh/fpga: 100
+  allocatable:
+    koordinator.sh/rdma: 100
+    koordinator.sh/fpga: 100
+```
+
+Why do we need `koordinator.sh/gpu-memory-ratio` and `koordinator.sh/gpu-memory` ? 
+When user apply 0.5/0.25 GPU, the user don't know the exact memory total bytes per GPU, only wants to use 
+half or quarter percentage of memory, so user can request the GPU memory with `koordinator.sh/gpu-memory-ratio`. 
+When scheduler assigned Pod on concrete node, scheduler will translate the `koordinator.sh/gpu-memory-ratio` to `koordinator.sh/gpu-memory` by the formulas:  ***allocatedMemory = totalMemoryOf(GPU)  * `koordinator.sh/gpu-memory-ratio`***, so that the GPU isolation can work.
```

Comment:
> scheduler will translate the `koordinator.sh/gpu-memory-ratio` to `koordinator.sh/gpu-memory`

Does this mean that scheduler will call kube-apiserver to update the pod's spec?

Reply:
it will update container resources.